container/sagemaker/python_service.py

```diff
@@ -0,0 +1,151 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+import logging
+import re
+from collections import namedtuple
+
+import falcon
+import requests
+
+logging.basicConfig(level=logging.INFO)
+log = logging.getLogger(__name__)
+
+TFS_BASE_URI = 'http://localhost:9001/v1/models/'
+TFS_DEFAULT_MODEL = 'hal_plus_three'
+TFS_DEFAULT_METHOD = 'predict'
+DEFAULT_CONTENT_TYPE = 'application/json'
+DEFAULT_ACCEPT_HEADER = 'application/json'
+
+Context = namedtuple('Context',
+                     'model_name, model_version, method, rest_uri, grpc_uri, '
+                     'custom_attributes, request_content_type, accept_header')
+
+
+class InvocationResource(object):
+
+    def __init__(self):
+        self.handler = None
+        self.input_handler = None
+        self.output_handler = None
+        self._import_handlers()
+
+    def on_post(self, req, res):
+        context, data = self._parse_http_request(req)
+
+        if self.handler:
+            log.warning('Using handler, input_handler and output_handler will be ignored.')
+            log.info('handler type: ' + str(type(self.handler)))
+            res.status = falcon.HTTP_200
+            res.body = self.handler(data, context)
+        elif self.input_handler and self.output_handler:
+            processed_input = self.input_handler(data, context)
+            response = requests.post(context.rest_uri, data=processed_input)
+            res.status = falcon.HTTP_200
+            res.body, res.content_type = self.output_handler(response, context)
+        else:
+            res.status = falcon.HTTP_500
+            res.body = 'Handlers are not implemented correctly.'
+
+    def _import_handlers(self):
+        # import handlers
+        import importlib.util
+        spec = importlib.util.spec_from_file_location('inference',
+                                                      '/opt/ml/model/inference_example.py')
+        inference = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(inference)
+
+        self.handler = inference.handler
+        self.input_handler = inference.input_handler
+        self.output_handler = inference.output_handler
+
+    def _parse_http_request(self, req):
+        content_type = req.get_header('Content-Type') or DEFAULT_CONTENT_TYPE
+        accept = req.get_header('Accept') or DEFAULT_ACCEPT_HEADER
+        custom_attributes = req.get_header('X-Amzn-SageMaker-Custom-Attributes')
+        model_name = req.get_header('model_name') or TFS_DEFAULT_MODEL
```

Comment:
how are these headers populated? usually we expect these values to arrive in the the 'X-Amzn-SageMaker-Custom-Attributes' header.

Reply:
I'm adding this `model_name` here just in case customer included such header in the request. It will be removed if model_name is specified in 'X-Amzn-SageMaker-Custom-Attributes', as we discussed in the comments of the design doc.

Reply:
I misunderstood that customer might want to set these headers (model_name, model_version ...). I should populate the context.xxx from 'X-Amzn-SageMaker-Custom-Attributes'

Reply:
yes, the model name, version, method all come from the `X-Amzn-SageMaker-Custom-Attributes` header in the invoke_endpoint request (or if the header is missing, default values are used). The user might also have put other values in this header, depending on the needs of their application.